src/kdl/serializable.cr

```diff
@@ -0,0 +1,308 @@
+            %var{name} = node.child({{name.stringify}}).properties.transform_values { |v, _| convert(v.value, {{ value[:type].type_vars[1] }}) }
+          {% else %}
+            %var{name} = {{value[:type]}}.from_kdl(node.child({{name.stringify}}))
+          {% end %}
+          %found{name} = true
+        {% end %}
+        {% for name, value in children_annos %}
+          # children
+          %var{name} = node.children.select { |n| n.name == {{value[:name]}} }.map { |n| {{value[:type].type_vars[0]}}.from_kdl(n) }
+          %found{name} = true
+        {% end %}
+
+        {% for name, value in all_properties %}
+          if %found{name}
+            @{{name}} = %var{name}
+          else
+            {% unless value[:has_default] || value[:nilable] %}
+              raise ::KDL::Serializable::Error.new("Missing KDL entry: {{value[:key].id}}", self.class.to_s, "{{ name }}")
+            {% end %}
+          end
+
+          {% if value[:presence] %}
+            @{{name}}_present = %found{name}
+          {% end %}
+        {% end %}
+      {% end %}
+    end
+
+    def to_kdl(node_name : String = self.class.to_s, builder : ::KDL::Builder = ::KDL::Builder.new)
+      {% begin %}
+        builder.node(node_name) do
+          {% for ivar in @type.instance_vars %}
+            {% name = ivar.id %}
+            {% if ann = ivar.annotation(::KDL::Argument) %}
+              builder.arg(@{{name}})
+            {% elsif ann = ivar.annotation(::KDL::Arguments) %}
+              @{{name}}.each do |arg|
+                builder.arg(arg)
+              end
+            {% elsif ann = ivar.annotation(::KDL::Property) %}
+              builder.prop({{(ann[:name] || ivar).id.stringify}}, @{{name}})
+            {% elsif ann = ivar.annotation(::KDL::Properties) %}
+              @{{name}}.each do |key, value|
+                builder.prop(key, value)
+              end
+            {% elsif ann = ivar.annotation(::KDL::Child) %}
+              {% if ann[:unwrap] == "argument" %}
+                builder.node({{(ann[:name] || ivar).id.stringify}}, @{{name}})
+              {% elsif ann[:unwrap] == "arguments" %}
+                builder.node({{(ann[:name] || ivar).id.stringify}}) do
+                  @{{name}}.each do |arg|
+                    builder.arg(arg)
+                  end
+                end
+              {% elsif ann[:unwrap] == "properties" %}
+                builder.node({{(ann[:name] || ivar).id.stringify}}, @{{name}})
+              {% else %}
+                @{{name}}.to_kdl({{(ann[:name] || ivar).id.stringify}}, builder)
+              {% end %}
+            {% elsif ann = ivar.annotation(::KDL::Children) %}
+              @{{name}}.each do |value|
+                value.to_kdl({{(ann[:name] || ivar).id.stringify}}, builder)
+              end
+            {% end %}
+          {% end %}
+        end
+      {% end %}
+    end
+  end
+end
+
+def Object.from_kdl(doc)
+  new doc
+end
```

Comment:
This should not be global, but should be in the module instead, as we can only call `#from_kdl` on Types that include `::KDL::Serializable`